### DISCUSSION TITLE
`xrdp_channel_process failed` error with `chansrvport=DISPLAY(0)`

### DISCUSSION THREAD

**Discussion by es-fabricemarie** [neutrinolabs/xrdp#3237] [Q&A]
I'm trying to get xrdp to connect statically to a x0vncserver session running on the main Xorg process. The goal is for a single user to connect via xrdp to the console running sddm, without having to enter a password in xrdp dialog.

The console and xrdp should show the same screen in the same state.

I'm using xrdp version 0.10.1 on Fedora 40.

My machine has the standard systemd sddm-x11 service running on the machine, where it starts an Xorg instance and awaits user login. I've made a systemd service to start x0vncserver for that display (after sddm service has started), as well as run xrdp-chansrv right after that. I've configured xrdp.ini to statically use that session.

The user can indeed connect and get the sddm screen after launching an RDP client and pressing enter on the xrdp login dialog.

The user can then proceed to login with sddm and and get a normal KDE session that just works.

My problem is with chansrv, xrdp doesn't seem to manage to connect to it. So all the channel operations like sound redirection, usb redirection etc do not work. When I run `lsof` to inspect the xrdp-chansrv process, I don't see any Unix sockets (that I can usually see in a more standard configuration of xrdp)

Am I missing some configuration or parameter when I start xrdp-chansrv?

The service I use to start x0vncserver is

```
[Unit]
Description=Remote desktop service (VNC) for :0 display
Requires=display-manager.service
After=network-online.target
After=display-manager.service

[Service]
Type=simple
Environment=HOME=/root
ExecStart=/usr/libexec/start_vnc0
Restart=on-failure
RestartSec=500ms

[Install]
WantedBy=multi-user.target
```

The file `/usr/libexec/start_vnc0` contains:

``` shell
XAUTHORITY=$(find /run/sddm/ -type f -name 'xauth_*')
export XAUTHORITY
export DISPLAY=:0
/usr/bin/x0vncserver \
    --display=:0 \
    --rfbauth=/etc/vncpasswd \
    --rfbport=5900 \
    --localhost \
    --AlwaysShared &
sleep 3
/usr/sbin/xrdp-chansrv
```

My `/etc/xrdp/xrdp.ini` contains:

``` ini
[Globals]
ini_version=1
fork=true
port=3389
use_vsock=false
tcp_nodelay=true
tcp_keepalive=true
security_layer=tls
crypt_level=high
certificate=
key_file=
ssl_protocols=TLSv1.2, TLSv1.3
autorun=
allow_channels=true
allow_multimon=true
bitmap_cache=true
bitmap_compression=true
bulk_compression=true
max_bpp=32
new_cursors=true
use_fastpath=both
grey=e1e1e1
dark_grey=b4b4b4
blue=0078d7
dark_blue=0078d7
ls_top_window_bg_color=003057
ls_width=350
ls_height=360
ls_bg_color=f0f0f0
ls_logo_filename=/usr/lib/evren_first_boot/evren_logo.svg
ls_logo_transform=scale
ls_logo_width=110
ls_logo_height=110
ls_logo_x_pos=135
ls_logo_y_pos=35
ls_label_x_pos=30
ls_label_width=68
ls_input_x_pos=110
ls_input_width=210
ls_input_y_pos=158
ls_btn_ok_x_pos=142
ls_btn_ok_y_pos=308
ls_btn_ok_width=85
ls_btn_ok_height=30
ls_btn_cancel_x_pos=237
ls_btn_cancel_y_pos=308
ls_btn_cancel_width=85
ls_btn_cancel_height=30
ls_title=Login to FAB-0012

[Logging]
LogFile=xrdp.log
LogLevel=INFO
EnableSyslog=true

[Channels]
rdpdr=true
rdpsnd=true
drdynvc=true
cliprdr=true
rail=true
xrdpvr=true

[Xvnc]
name=default
lib=libvnc.so
username=
password=password
ip=127.0.0.1
port=5900
chansrvport=DISPLAY(0)
```

My `/etc/xrdp/sesman.ini` file is:

``` ini
[Globals]
EnableUserWindowManager=true
UserWindowManager=startwm.sh
DefaultWindowManager=startwm-bash.sh
ReconnectScript=reconnectwm.sh

[Security]
AllowRootLogin=true
MaxLoginRetry=4
TerminalServerUsers=tsusers
TerminalServerAdmins=tsadmins
AlwaysGroupCheck=false
RestrictOutboundClipboard=none
RestrictInboundClipboard=none

[Sessions]
X11DisplayOffset=10
MaxSessions=50
KillDisconnected=true
DisconnectedTimeLimit=60
IdleTimeLimit=0
Policy=Default

[Logging]
LogFile=xrdp-sesman.log
LogLevel=INFO
EnableSyslog=false

[Xorg]
param=/usr/libexec/Xorg
param=-config
param=xrdp/xorg.conf
param=-noreset
param=-nolisten
param=tcp
param=-logfile
param=.xorgxrdp.%s.log

[Xvnc]
param=Xvnc
param=-bs
param=-nolisten
param=tcp
param=-localhost
param=-dpi
param=96

[Chansrv]
FuseMountName=thinclient_drives
FileUmask=077

[ChansrvLogging]
LogLevel=DEBUG
EnableSyslog=false
EnableConsole=true
ConsoleLevel=DEBUG

[SessionVariables]
PULSE_SCRIPT=/etc/xrdp/pulse/default.pa
```

sesman log file is almost empty and doesn't look relevant here.
xrdp log file is:

```
[INFO ] starting xrdp with pid 375113
[INFO ] address [0.0.0.0] port [3389] mode 1
[INFO ] listening to port 3389 on 0.0.0.0
[INFO ] xrdp_listen_pp done
[INFO ] Socket 14: connection accepted from [::ffff:192.168.122.1]:49796
[INFO ] Using default X.509 certificate: /etc/xrdp/cert.pem
[INFO ] Using default X.509 key file: /etc/xrdp/key.pem
[INFO ] Security protocol: configured [SSL], requested [SSL|HYBRID|RDP], selected [SSL]
[INFO ] Connected client computer name: FAB-0001
[INFO ] client supports gfx protocol
[WARN ] Physical desktop dimensions (0x0) are invalid
[WARN ] Received [MS-RDPBCGR] TS_UD_HEADER type 0xc006 is unknown (ignored)
[WARN ] Received [MS-RDPBCGR] TS_UD_HEADER type 0xc00a is unknown (ignored)
[INFO ] xrdp_load_keyboard_layout: Keyboard information sent by the RDP client, keyboard_type:[0x04], keyboard_subtype:[0x00], keylayout:[0x00020409]
[INFO ] xrdp_load_keyboard_layout: model [] variant [] layout [us] options []
[INFO ] TLS connection established from [::ffff:192.168.122.1]:49796 TLSv1.3 with cipher TLS_AES_256_GCM_SHA384
[INFO ] Client supports single-screen resizes by xrdp
[INFO ] xrdp_caps_process_pointer: client supports new(color) cursor
[INFO ] xrdp_caps_process_codecs: RemoteFX(76772F12-BD72-4463-AFB3-B73C9C6F7886), codec id [3], properties len [49]
[INFO ] Client Capability: LARGE_POINTER_FLAG_96x96 supported
[WARN ] Cannot find keymap file /etc/xrdp/km-00020409.ini
[INFO ] Loading keymap file /etc/xrdp/km-00000409.ini
[WARN ] local keymap file for 0x00020409 found and doesn't match built in keymap, using local keymap file
[WARN ] No information is available to determine login screen DPI
[WARN ] No DPI value is available to find login font
[WARN ] Using the default_dpi of 96
[INFO ] xrdp_egfx_create: error 0 channel_id 1
[INFO ] xrdp_mm_egfx_caps_advertise:
[INFO ]   version 0x00080004 flags 0x00000001 (index: 0)
[INFO ]   version 0x00080105 flags 0x00000013 (index: 1)
[INFO ]   version 0x000a0002 flags 0x00000002 (index: 2)
[INFO ]   version 0x000a0200 flags 0x00000002 (index: 3)
[INFO ]   version 0x000a0301 flags 0x00000040 (index: 4)
[INFO ]   version 0x000a0400 flags 0x00000042 (index: 5)
[INFO ]   version 0x000a0502 flags 0x00000042 (index: 6)
[INFO ]   version 0x000a0600 flags 0x00000042 (index: 7)
[INFO ]   version 0x000a0601 flags 0x00000042 (index: 8)
[INFO ] unknown version 0x000a0601
[INFO ]   version 0x000a0701 flags 0x000000c2 (index: 9)
[INFO ]   replying version 0x000a0701 flags 0x000000c2
[INFO ] xrdp_mm_egfx_caps_advertise: xrdp_egfx_send_capsconfirm error 0 best_index 9
[INFO ] xrdp_egfx_reset_graphics:
[INFO ] xrdp_egfx_reset_graphics: width 1280 height 800 monitorcount 1
[INFO ] xrdp_mm_egfx_caps_advertise: xrdp_egfx_send_reset_graphics error 0 monitorCount 0
[INFO ] xrdp_mm_egfx_create_surfaces: map surface_id 0 left 0 top 0 width 1280 height 800
[INFO ] xrdp_encoder_create: starting gfx rfx pro codec session
[INFO ] xrdp_mm_egfx_invalidate_wm_screen:
[INFO ] xrdp_mm_egfx_caps_advertise: egfx created.
[ERROR] dynamic_monitor_open_response: error
[ERROR] xrdp_rdp_recv: xrdp_channel_process failed
[INFO ] xrdp_wm_log_msg: Connecting to session
[INFO ] loaded module 'libvnc.so' ok, interface size 4832, version 4
[INFO ] VNC: Clipboard (if available) is provided by chansrv facility
[INFO ] xrdp_wm_log_msg: Connecting to chansrv
[ERROR] xrdp_mm_chansrv_connect: error in trans_connect chan
[INFO ] Layout from ClientLayout (geom=1280x800 #screens=1) : 0:(1280x800+0+0)
[INFO ] Layout from OldServerLayout (geom=1280x800 #screens=1) : 1804289383:(1280x800+0+0)
```

After I disconnect the session, xrdp.log adds these lines:

```
[ERROR] xrdp_iso_send: trans_write_copy_s failed
[ERROR] Sending [ITU T.125] DisconnectProviderUltimatum failed
[ERROR] xrdp_iso_send: trans_write_copy_s failed
[ERROR] xrdp_mcs_send: xrdp_iso_send failed
[ERROR] xrdp_sec_send: xrdp_mcs_send failed
[ERROR] xrdp_channel_send: xrdp_sec_send failed
[ERROR] xrdp_channel_drdynvc_data: xrdp_channel_send failed
```

I was following the hints from https://github.com/neutrinolabs/xrdp/issues/1891#issuecomment-844931523
given by @matt335672 regarding `chansrvport=DISPLAY(0)` but I seem to be missing something.

**Comment by matt335672** [comment]
@es-fabricemarie 

This has been raised in a comment before by a user @usa- who now appears to have disappeared from Github.

Most likely this is a regression. I'll try to get round to looking at it later today.

**Reply by iGraveD** [reply]
User @usa- asked me to tell you, that he hasn't disappeared, he sees everything, it's just that GitHub flagged his account for some unknown reason. That's why you temporarily (hopefully) can't see the tickets he created.

**Comment by matt335672** [comment]
@es-fabricemarie - one thing springs to mind. Can you check your display number?

If you've configured gdm on your machine for X operation only, display :0 will be used for gdm, and another display (probably :1) will be used for your session.

Check your display number with `echo $DISPLAY`

You may need to set `chansrvport=DISPLAY(1)` (or similar)

**Reply by es-fabricemarie** [reply]
Thanks for the help!

I'm using sddm and not gdm.

Yes the display number :0 seems correct as I do manage to get the sddm login window through RDP: I simply click ok on xrdp login window and xrdp shows me sddm login window. From there I can login as per usual.

My issue is only with xrdp-chansrv.

x0vncserver is started as root right after sddm, and xrdp-chansrv as root right after x0vncserver with the same DISPLAY and XAUTHORITY env variables.

Does xrdp-chansrv require additional environment variables before being launched like that?

**Comment by matt335672** [comment]
xrdp-chansrv needs to run as the same user as the X session.

If you run it as root, `xrdp` won't be able to find the socket for it.

**Reply by es-fabricemarie** [reply]
I'm trying to get xrdp to act like a console basically: to let the user interact with sddm and the session that the user creates by logging in with sddm. xrdp-sesman gets completely bypassed here.

Both my Xorg and my sddm run as root:

```
root      435598  0.0  0.5 375812 23624 ?        Ssl  16:57   0:00 /usr/bin/sddm
root      435607  0.0  3.0 635268 120348 tty1    Ssl+ 16:57   0:11  \_ /usr/libexec/Xorg -nolisten tcp -background none -seat seat0 vt1 -auth /run/sddm/xauth_JmpfUQ -noreset -displayfd 15
```

There is no user session at that stage as user has not logged-in through sddm yet.

I did an `lsof` of xrdp-chansrv (when started as root with correct XAUTHORITY / DISPLAY) but couldn't see any Unix socket at all.

From what I see, it looks like xrdp-chansrv gets started by xrdp-sesman, but since my use-case bypasses totally sesman it looks like xrdp-chansrv is not happy.

If I can't start xrdp-chansrv as root (before the user gets his/her session), can I start xrdp-chanserv as the end-user (with an XDG autostart `.desktop` file for example)? Would xrdp detect chansrv has been started and start using it?

**Reply by matt335672** [reply]
OK - I can see what's happening here. It's to do with the sesman bypass.

chansrv creates UNIX domain sockets to communicate with xrdp. See this page for details:-

https://github.com/neutrinolabs/xrdp/wiki/The-socketdir-directory

When the session is created by sesman, sesman (running as root) creates a directory for these sockets to be stored in. The directory is only accessible by `xrdp` and the user to prevent any information leakage.

chansrv expects this directory to be present. If it isn't, it can't create the sockets.

Please have a look for the chansrv log file (normally in `$HOME/.local/xrdp/xrdp-chansrv.0.log`) and report back. Other configurations will be affected by this, and it's not clear to me what a secure solution would look like yet.

**Reply by matt335672** [reply]
Sorry, that's `$HOME/.local/share/xrdp/xrdp-chansrv.0.log`

I've opened #3240 on this to gather more information.

**Reply by es-fabricemarie** [reply]
Thanks for the detailed explanation, this helps a lot.

The directory exists and is populated (by chanserv) properly:

```
find /run/xrdp/0 -ls
     6993      0 drwxr-s---   2 root     root           80 Sep 10 00:46 /run/xrdp/0
    11549      0 srw-rw----   1 root     root            0 Sep 10 00:46 /run/xrdp/0/xrdpapi_0
    11550      0 srw-rw----   1 root     root            0 Sep 10 00:46 /run/xrdp/0/xrdp_chansrv_socket_0
```

I've tried to not start chanserv at all, and in that case the xrdp session simply fails completely. So it means chanserv seem to at least start properly.

chanserv log file just says it closed the socket it opened earlier on:

```
[2024-09-10T00:52:43.336+0800] [DEBUG] Closed socket 13 (AF_UNIX)
[2024-09-10T00:52:43.509+0800] [DEBUG] Closed socket 12 (AF_UNIX)
```

When I try on another machine where I configured xrdp in completely standard way with xvnc and dynamic sessions, I could see these files with lsof:

```
/run/xrdp/1001/xrdpapi_10
/run/xrdp/1001/xrdp_chansrv_socket_10
/run/xrdp/1001/xrdp_chansrv_audio_out_socket_10
/run/xrdp/1001/xrdp_chansrv_audio_in_socket_10
/home/fabrice/.pcsc10/pcscd.comm
```

Could the error messages I was seeing earlier on be related to sound?
I'm on Fedora with pipewire. (sound work perfectly with https://github.com/neutrinolabs/pipewire-module-xrdp when configured with xvnc and dynamic sessions, no worries).

I'll try to simulate what https://github.com/neutrinolabs/pipewire-module-xrdp/blob/devel/instfiles/load_pw_modules.sh script does but at startup time since I want a static session.

I'll get back shortly with the result.

**Reply by matt335672** [reply]
I had a look at the code yesterday, and I'm not convinced that `chansrvport=DISPLAY(0)` is working in all situations. It needs to understand the concept of the current user (or UID), and that's not very well defined for your configuration.

1) Can you post the stanza you're using in `xrdp.ini` so I can step through it?
2) Try replacing it with `chansrvport=/run/xrdp/0/xrdp_chansrv_socket_0` for now.

As for the sound channels, these are only created when xrdp sends the channel list through from the client. Since we're not getting that connection, the channels won't be set up.

**Reply by es-fabricemarie** [reply]
Thanks @matt335672 , this is progressing very well.

2. My stanza for xrdp.ini is in my original post. Sorry, the post was a bit big and messy.
3. Your point number 2. above (`chansrvport=/run/xrdp/0/xrdp_chansrv_socket_0`) goes a long way. I now can see:
    ```
    find /run/xrdp/0 -ls
         4054      0 drwxr-xr-x   2 root     root          100 Sep 10 19:57 /run/xrdp/0
         5183      0 srw-rw----   1 root     root            0 Sep 10 19:56 /run/xrdp/0/xrdpapi_0
         5193      0 srw-rw----   1 root     root            0 Sep 10 19:57 /run/xrdp/0/xrdp_chansrv_audio_out_socket_0
         5194      0 srw-rw----   1 root     root            0 Sep 10 19:57 /run/xrdp/0/xrdp_chansrv_audio_in_socket_0
    ```
Let me fully disable spice (which might be hijacking the sound before xrdp does) and try again.

If it doesn't work, I'll peer through all the logs and check what could be the issue.

**Reply by es-fabricemarie** [reply]
I still get a tiny error (not sure if it's relevant), when I connect the RDP client to xrdp (but don't click ok yet in the xrdp blue screen to get the user session):

```
[INFO ] Socket 14: connection accepted from [::ffff:192.168.122.1]:50994
[INFO ] Using default X.509 certificate: /etc/xrdp/cert.pem
[INFO ] Using default X.509 key file: /etc/xrdp/key.pem
[INFO ] Security protocol: configured [SSL], requested [SSL|HYBRID|RDP], selected [SSL]
[INFO ] Connected client computer name: FAB-0001
[INFO ] client supports gfx protocol
[WARN ] Physical desktop dimensions (0x0) are invalid
[WARN ] Received [MS-RDPBCGR] TS_UD_HEADER type 0xc006 is unknown (ignored)
[WARN ] Received [MS-RDPBCGR] TS_UD_HEADER type 0xc00a is unknown (ignored)
[INFO ] xrdp_load_keyboard_layout: Keyboard information sent by the RDP client, keyboard_type:[0x04], keyboard_subtype:[0x00], keylayout:[0x00020409]
[INFO ] xrdp_load_keyboard_layout: model [] variant [] layout [us] options []
[INFO ] TLS connection established from [::ffff:192.168.122.1]:50994 TLSv1.3 with cipher TLS_AES_256_GCM_SHA384
[INFO ] Client supports single-screen resizes by xrdp
[INFO ] xrdp_caps_process_pointer: client supports new(color) cursor
[INFO ] xrdp_caps_process_codecs: RemoteFX(76772F12-BD72-4463-AFB3-B73C9C6F7886), codec id [3], properties len [49]
[INFO ] Client Capability: LARGE_POINTER_FLAG_96x96 supported
[WARN ] Cannot find keymap file /etc/xrdp/km-00020409.ini
[INFO ] Loading keymap file /etc/xrdp/km-00000409.ini
[WARN ] local keymap file for 0x00020409 found and doesn't match built in keymap, using local keymap file
[WARN ] No information is available to determine login screen DPI
[WARN ] No DPI value is available to find login font
[WARN ] Using the default_dpi of 96
[INFO ] xrdp_egfx_create: error 0 channel_id 1
[INFO ] xrdp_mm_egfx_caps_advertise:
[INFO ]   version 0x00080004 flags 0x00000001 (index: 0)
[INFO ]   version 0x00080105 flags 0x00000013 (index: 1)
[INFO ]   version 0x000a0002 flags 0x00000002 (index: 2)
[INFO ]   version 0x000a0200 flags 0x00000002 (index: 3)
[INFO ]   version 0x000a0301 flags 0x00000040 (index: 4)
[INFO ]   version 0x000a0400 flags 0x00000042 (index: 5)
[INFO ]   version 0x000a0502 flags 0x00000042 (index: 6)
[INFO ]   version 0x000a0600 flags 0x00000042 (index: 7)
[INFO ]   version 0x000a0601 flags 0x00000042 (index: 8)
[INFO ] unknown version 0x000a0601
[INFO ]   version 0x000a0701 flags 0x000000c2 (index: 9)
[INFO ]   replying version 0x000a0701 flags 0x000000c2
[INFO ] xrdp_mm_egfx_caps_advertise: xrdp_egfx_send_capsconfirm error 0 best_index 9
[INFO ] xrdp_egfx_reset_graphics:
[INFO ] xrdp_egfx_reset_graphics: width 1440 height 900 monitorcount 1
[INFO ] xrdp_mm_egfx_caps_advertise: xrdp_egfx_send_reset_graphics error 0 monitorCount 0
[INFO ] xrdp_mm_egfx_create_surfaces: map surface_id 0 left 0 top 0 width 1440 height 900
[INFO ] xrdp_encoder_create: starting gfx rfx pro codec session
[INFO ] xrdp_mm_egfx_invalidate_wm_screen:
[INFO ] xrdp_mm_egfx_caps_advertise: egfx created.
[ERROR] dynamic_monitor_open_response: error
[ERROR] xrdp_rdp_recv: xrdp_channel_process failed
<blue login screen appears in xrdp asking me to click OK>
```

This could be normal since I'm connecting to x0vncserver which is backed by Xorg on my real screen, I'm not sure.

Before this last iteration, after I clicked "ok" to connect the session, I would get these logs:

```
[INFO ] xrdp_wm_log_msg: Connecting to chansrv
[ERROR] xrdp_mm_chansrv_connect: error in trans_connect chan
```

But now after making the `chansrvport=/run/xrdp/0/xrdp_chansrv_socket_0` change, this last error disapeared.

**Comment by es-fabricemarie** [comment]
I managed to make it work in the end.

I've added a `/etc/profile.d/xrdp_static.sh` file

```shell
export XRDP_SOCKET_PATH=/run/xrdp/0
export XRDP_PULSE_SOURCE_SOCKET=xrdp_chansrv_audio_in_socket_0
export XRDP_PULSE_SINK_SOCKET=xrdp_chansrv_audio_out_socket_0
export PULSE_SCRIPT=/etc/xrdp/pulse/default.pa
export XRDP_SESSION="blah"
```

and I've changed my `/usr/libexec/start_vnc0` file with:

```shell
#!/bin/bash
XAUTHORITY=$(find /run/sddm/ -type f -name 'xauth_*')
export XAUTHORITY
export DISPLAY=:0
export USER=root
export HOME=/root
export XDG_CONFIG_DIRS=/root/.config/kdedefaults:/etc/xdg:/usr/share/kde-settings/kde-profile/default/xdg
export XDG_DATA_HOME=/root/.local/share
export LOGNAME=root
export KDE_SESSION_UID=0
export XDG_RUNTIME_DIR=/run/user/0
export XRDP_SESSION=1
sleep 2
# start the VNC server
/usr/bin/x0vncserver \
    --display=:0 \
    --rfbauth=/etc/entsecure/vncpasswd \
    --rfbport=5900 \
    --localhost \
    --AlwaysShared &
# ensure all your users are part of the "users" group, or choose the correct one.
# create the directory in advance and add set-gid to the directory
# so regular users can connect
mkdir -p /run/xrdp/0
chown root:users /run/xrdp/0
chmod g+s /run/xrdp/0
/usr/sbin/xrdp-chansrv
```

Finally, I deleted the `/etc/xdg/autostart/xrdp-chansrv.desktop` file as in my case chansrv is started statically first.

I haven't tried copy/paste and USB redirections yet, but sounds work.

**Reply by matt335672** [reply]
That's great. If sound is working you should be there for the other chansrv functions.

The error is coming about as xrdp is trying to open a channel to the client to get dynamic updates. This is clearly pointless in your configuration. Can you add the following to your `[Xvnc]` stanza and see if the error goes away for you:-

```
enable_dynamic_resizing=false
```

Sorry I missed your stanza earlier - I can see it now:-

```ini
[Xvnc]
name=default
lib=libvnc.so
username=
password=password
ip=127.0.0.1
port=5900
chansrvport=DISPLAY(0)
```

In the move to v0.10.x we improved the socketdir security quite a bit, Consequently, the `DISPLAY()` mechanism needs access to a UID from somewhere to work properly. In v0.10.x this is done by asking sesman to authenticate a user somehow as this is when the username to UID translation is performed.

In your somewhat unusual setup, there's no user and no authentication, so the DISPLAY() mechanism is silently failing. This needs improvement. My current thinking is:-
1) Fix the logging!
2) Add an optional UID to the mechanism so you could use DISPLAY(0,0) for instance.

I also need to give chansrv a way to create the socket dir if it doesn't exist, (although this doesn't apply to your setup). Because of the permissions involved, it's likely chansrv will need to ask sesman to do this.

I'd appreciate any thoughts you have on this, particularly as you're coming at it from a completely different angle - I've probably missed something else somewhere.

**Reply by es-fabricemarie** [reply]
This indeed removes the error, thanks:

```
enable_dynamic_resizing=false
```

I agree with your recommendations on logging and directory creation, yes.

My main issue in a nutshell is that since I'm doing xrdp before user authentication (and don't use sesman+rdp+pam for auth) my chansrv process has to be owned by root.

chansrv doesn't give me issues with sound redirection, as I handle the file permission of the xrdp socket dir and grant group permission. But it does give issues with other redirections. In my current config for example, drives "sharing" from windows will mount the drives over at:

```
/root/thinclient_drives/C:
/root/thinclient_drives/.clipboard
/root/thinclient_drives/D:
/root/thinclient_drives/F:
```
And all the files in there are owned by root. Which makes it impossible for a non-privileged user to access.

I've since enabled vsocket in the xrdp.ini config file and made the daemon listen to vsocket and tcp:

```
port=vsock://-1:3389 tcp://:3389
use_vsock=true
```

I'm now seeing this error when I test the redirections further:

```
Sep 15 22:41:32 FAB-0012 xrdp[10509]: [INFO ] Socket 15: connection accepted from 172.21.208.1:64889
Sep 15 22:41:34 FAB-0012 xrdp[16126]: [INFO ] Using default X.509 certificate: /etc/xrdp/cert.pem
Sep 15 22:41:34 FAB-0012 xrdp[16126]: [INFO ] Using default X.509 key file: /etc/xrdp/key.pem
Sep 15 22:41:38 FAB-0012 xrdp[16126]: [INFO ] Security protocol: configured [SSL], requested [SSL|HYBRID|HYBRID_EX|RDP], selected [SSL]
Sep 15 22:41:38 FAB-0012 xrdp[10509]: [INFO ] Socket 15: connection accepted from 172.21.208.1:64890
Sep 15 22:41:39 FAB-0012 xrdp[16126]: [ERROR] SSL_read: I/O error
Sep 15 22:41:39 FAB-0012 xrdp[16126]: [ERROR] libxrdp_force_read: header read error
Sep 15 22:41:39 FAB-0012 xrdp[16126]: [ERROR] Processing [ITU-T T.125] Connect-Initial failed
Sep 15 22:41:39 FAB-0012 xrdp[16126]: [ERROR] [MCS Connection Sequence] receive connection request failed
Sep 15 22:41:40 FAB-0012 xrdp[16126]: [ERROR] xrdp_sec_incoming: xrdp_mcs_incoming failed
Sep 15 22:41:40 FAB-0012 xrdp[16126]: [ERROR] xrdp_rdp_incoming: xrdp_sec_incoming failed
Sep 15 22:41:40 FAB-0012 xrdp[16126]: [ERROR] xrdp_process_main_loop: libxrdp_process_incoming failed
Sep 15 22:41:40 FAB-0012 xrdp[16126]: [ERROR] xrdp_iso_send: trans_write_copy_s failed
Sep 15 22:41:41 FAB-0012 xrdp[16126]: [ERROR] Sending [ITU T.125] DisconnectProviderUltimatum failed
Sep 15 22:41:41 FAB-0012 xrdp[16135]: [INFO ] Using default X.509 certificate: /etc/xrdp/cert.pem
Sep 15 22:41:41 FAB-0012 xrdp[16135]: [INFO ] Using default X.509 key file: /etc/xrdp/key.pem
```

Which I have no idea how to interpret for now. But I'm happy to help debug them.

To solve the issues I'm facing, I was thinking if there were a way that I could run the main chansrv process as root as I do now, but start a sub-chansrv process from the non-privileged user's session (for example with xdg autostart mechanism) in a way that they talk to each other?

I don't manage to make it work right now as there seem to be some shared secret/knowledge between the two? I've noticed there is a `XRDP_SESSION` environment variable for example, and it points to the sesman PID. But I'm not using sesman...

Any idea if that is possible?

**Reply by matt335672** [reply]
I don't know if the kernel will let you share FUSE devices in this way, so that may not be possible.

However, you can play with these options in sesman.ini and see if they can solve this for you:-

1) `FuseMountName`
2) `FileUmask`

The errors you list are (surprisingly) perfectly normal. See #2040 for the background.

We've got no provision for chaining chansrv processes, but it's an interesting idea! It would be horribly slow however, and I've no idea how we could make the permissions work in a sane way.

`XRDP_SESSION`  should be set to the same PID as the `sesexec` process, not `sesman`. Can you check that? If it's not there's something else strange going on in your config.

**Reply by es-fabricemarie** [reply]
`FuseMountName` and `FileUmask` yes, I just just just discovered it in the man-page of sesman.ini, thanks, I'll set them to a shared directory as only one user will ever connect at the same time as this is for an HyperV VM.

For the errors mentioned in 2040 thanks, that explains everything and makes a lot of sense.

`XRDP_SESSION` was indeed set to the PID of `sesexec` I remembered wrongly that's all. But only when I was using sesman to login to the system. Now with my setup neither sesman nor sesexec are used. What is `XRDP_SESSION` variable used for? It doesn't appear to be used on my system..?

What I meant was not really chaining chansrv processes, but more like take-over when there is an (external to sesman) user session started. So it would be something like this:

root chansrv running (no user logged in)
--> user logs in through RDP on sddm and connect to chansrv ran as root
--> user runs their own chansrv process, asking root chansrv how to re-setup the redirections or take them over from the main one.

Then when user logs out
--> user chansrv process tell main/root chansrv to take over as it's quitting (or root chansrv could detect that)
--> root chansrv takes over

That would complicate things a bit initially but makes a lot of sense on some setups (like mine)

**Reply by matt335672** [reply]
`XRDP_SESSION` is set by `sesexec` to allow a session to be identified. It's not used for a lot at present, but will be later on.

Let's crack on with seeing if we can solve your existing chansrv issues first with `sesman.ini` configuration.
